src/config/builder.rs

```diff
@@ -0,0 +1,150 @@
+use super::{
+    compiler, default_data_dir, Config, GlobalOptions, SinkConfig, SinkOuter, SourceConfig,
+    TestDefinition, TransformConfig, TransformOuter,
+};
+use crate::event;
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize, Debug, Default)]
+#[serde(deny_unknown_fields)]
+pub struct ConfigBuilder {
+    #[serde(flatten)]
+    pub global: GlobalOptions,
+    #[serde(default)]
+    pub sources: IndexMap<String, Box<dyn SourceConfig>>,
+    #[serde(default)]
+    pub sinks: IndexMap<String, SinkOuter>,
+    #[serde(default)]
+    pub transforms: IndexMap<String, TransformOuter>,
+    #[serde(default)]
+    pub tests: Vec<TestDefinition>,
+}
+
+impl Clone for ConfigBuilder {
+    fn clone(&self) -> Self {
+        // This is a hack around the issue of cloning
+        // trait objects. So instead to clone the config
+        // we first serialize it into JSON, then back from
+        // JSON. Originally we used TOML here but TOML does not
+        // support serializing `None`.
+        let json = serde_json::to_vec(self).unwrap();
```

Comment:
Could you turn it into a `serde_json::Value` instead to save on the cost of making everything into a contiguous byte array?